src/include/sof/schedule/dp_schedule.h

```diff
@@ -0,0 +1,99 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
+ *
+ * Author: Marcin Szkudlinski
+ */
+
+#ifndef __SOF_SCHEDULE_DP_SCHEDULE_H__
+#define __SOF_SCHEDULE_DP_SCHEDULE_H__
+
+#include <sof/schedule/task.h>
+#include <sof/trace/trace.h>
+#include <user/trace.h>
+#include <stdint.h>
+
+/**
+ *
+ * DP scheduler is a scheduler that creates a separate preemptible Zephyr thread for each SOF task
+ *   There's only one instance of DP in the system, however, the threads may run on any core
+ *   Threads are pinned to specific core, there's no SMP processing.
+ *
+ * The task execution may be delayed and task may be re-scheduler periodically
+ * NOTE: delayed start and rescheduling takes place in sync with LL scheduler, meaning the
+ *   DP scheduler is triggered as the last task of LL. That implicates a limitation: LL scheduler
+ *   MUST be running on primary core in order to have this feature working.
+ *   It is fine, because rescheduling is a feature used for data processing when a pipeline is
+ *   running.
+ *
+ * Other possible usage of DP scheduler is to schedule task with DP_SCHEDULER_RUN_TASK_IMMEDIATELY
+ *   as start parameter. It will force the task to work without any delays and asyn to LL.
+ *   This kind of scheduling may be used for tasks like IPC.
```

Comment:
1) Won't simplify much, couple of lines
2) in future there's a concept of "taksks with budgets", not supported by zephyr directly. We discuss it with Zephyr and they only provide some callback to make it possible to implement
3) I would keep an abstraction for a) porting b) common API with schedulers like LL working in completely diffrent way

Reply:
I meant tasks like IPC. IPC with Zephyr is currently using the EDF scheduler, which is just a thin API wrapper around Zephyr work queues, not much more than that.

Reply:
I know about current zephyr_EDF, but we need to have IPC to be preemptive also, as a "task with budget" - IPC should be processed very quickly, but if processing takes to long it must yield to LL. Current zephyr implementation of EDF, as you said, does not do much and is not a really EDF, just "start the earliest deadline task and hope it finish processing in time".

Reply:
@marcinszkudlinski @lyakh Let's not detour to IPC too much here. The current IPC with EDF is clearly not optimal and should be either written on top of native work queues, or something else if we need to have preemption after budget has expired (has not implemented so far in SOF, so this itself is a new feature).

Reply:
ok, I know is not implemented, but is scheduled (sic!) for implementation :)

Reply:
@lyakh, @kv2019i we detour to IPC for a reason. Works on entering D3 power state with Zephyr Power Manager had been suspended as there is a bug in zephyr k_waitable abstraction that handles scheduling ipc task for future (after wake from D3). No progress had been made on the subject so we are trying hard with @marcinszkudlinski to replace the scheduling system.

Reply:
@marcinszkudlinski @aborisovich @kv2019i why are you saying that the current IPC thread isn't preemptible? As long as it has a lower priority than the LL scheduler it will be preempted by it, regardless of any budgets. When the next timer interrupt triggers, interrupting the IPC thread, on exit the Zephyr scheduler will select a runnable task with the highest priority and that will be the LL scheduler thread.

Reply:
All tasks in current "edf" implementation is based on Zephyr work queue and run by a single Zephyr thread. Right, this thread can be preempted. BUT the tasks themselves are cooperative to each other. All the current edf can do is to decide which task from the queue starts first

Reply:
Indeed, the Zephyr work queue is an in-order list of items to run, it doesn't handle preemption between tasks (which aren't threads, so they can't **be** preempted, they're just callbacks that run on the same thread).   This isn't really a bug, it's just the way the data structure works.

If you want to preempt something, it has to be running in a thread with its own stack that can be suspended.

Note that there's an alternative, the "p4wq", which is a little more complicated, but based on the idea of running queue items in pooled threads such that they all have their own priorities and can preempt each other, but without the overhead of having to spawn one thread per task.  This was actually written with SOF in mind originally, as it happens, though it will likely require some work to enable.  See https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/p4wq.h

Reply:
@andyross maybe I misunderstood your "it will likely require some work to enable" but it is already used by SOF for IDC

Reply:
So, yes, as a further optimisation we could in the future consider moving this to p4wq, that would allow us to use threads more optimally - use a common pool of threads instead of a thread per task, right @andyross ?